vmm/src/vm.rs

```diff
@@ -718,14 +569,203 @@ impl<'a> Vm<'a> {
+        name: String,
+        virtio_device: Box<vm_virtio::VirtioDevice>,
+        memory: GuestMemoryMmap,
+        vm_fd: &VmFd,
+        pci: &mut Arc<Mutex<PciConfigIo>>,
+        device_manager: &mut DeviceManager,
+        msi_capable: bool,
+    ) -> Result<()> {
+        let msix_num = if msi_capable { DEFAULT_MSIX_VEC_NUM } else { 0 };
+
+        let virtio_pci_device = VirtioPciDevice::new(name.clone(), memory, virtio_device)
+            .map_err(Error::VirtioDevice)?;
+
+        // Register virtio device resource
+        let mut resource = virtio_pci_device.get_resource();
+
+        let virtio_pci_device = Arc::new(Mutex::new(virtio_pci_device));
+
+        device_manager
+            .register_device(
+                virtio_pci_device.clone(),
+                Some(pci.clone()),
+                &mut resource,
+                Some(IrqResource(None)),
+            )
+            .map_err(Error::RegisterDevice)?;
+
+        pci.lock()
+            .expect("failed to acquire lock")
+            .add_device(virtio_pci_device.clone())
+            .map_err(Error::AddPciDevice)?;
+
+        for (event, addr, _) in virtio_pci_device
+            .lock()
+            .expect("failed to unlock")
+            .ioeventfds()
+        {
+            let io_addr = IoEventAddress::Mmio(addr);
+            vm_fd
+                .register_ioevent(event.as_raw_fd(), &io_addr, NoDatamatch)
+                .map_err(Error::RegisterIoevent)?;
+        }
+
+        if msi_capable {
+            let vm_fd_clone = vm_fd.clone();
+
+            let msi_cb = Arc::new(Box::new(move |p: InterruptParameters| {
+                if let Some(entry) = p.msix {
+                    let msi_queue = kvm_msi {
+                        address_lo: entry.msg_addr_lo,
+                        address_hi: entry.msg_addr_hi,
+                        data: entry.msg_data,
+                        flags: 0u32,
+                        devid: 0u32,
+                        pad: [0u8; 12],
+                    };
+
+                    return vm_fd_clone.signal_msi(msi_queue).map(|ret| {
+                        if ret > 0 {
+                            debug!("MSI message successfully delivered");
+                        } else if ret == 0 {
+                            warn!("failed to deliver MSI message, blocked by guest");
+                        }
+                    });
+                }
+
+                Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    "missing MSI-X entry",
+                ))
+            }) as InterruptDelivery);
+
+            virtio_pci_device.assign_msix(msi_cb);
+        } else {
+            let irq_num = virtio_pci_device.lock().expect("failed").irq_num();
+            let event_fd = virtio_pci_device
+                .lock()
+                .expect("failed")
+                .interrupt_evt()
+                .unwrap()
+                .as_raw_fd();
+
+            vm_fd
```

Comment:
@sameo  Also here. The eventfd needs to be reused after setting it into callback.